File: backend/src/routes/tenders.js
import express from 'express';
import fs from 'fs';
import { prisma } from '../db.js';
import { get as cacheGet, set as cacheSet } from '../cache.js';
import { aiExtractTender } from '../pipeline/aiExtract.js';
import pdfParse from 'pdf-parse/lib/pdf-parse.js';

const router = express.Router();

const SORT_MAP = {
  endDate_asc:    { endDate: 'asc' },
  endDate_desc:   { endDate: 'desc' },
  bidValue_asc:   { bidValue: 'asc' },
  bidValue_desc:  { bidValue: 'desc' },
  emdAmount_asc:  { emdAmount: 'asc' },
  emdAmount_desc: { emdAmount: 'desc' },
  fetchedAt_desc: { fetchedAt: 'desc' },
};

/**
 * GET /api/tenders
 * Filters: city, q, category, status (open|closed|all, default open),
 * minValue, maxValue, minEmd, maxEmd, source (GEM|CSPGCL|all),
 * sort (default endDate_asc), page (default 1), limit (default 20, max 100)
 */
router.get('/', async (req, res) => {
  try {
    const cacheKey = 'tenders:' + JSON.stringify(req.query);
    const cached = cacheGet(cacheKey);
    if (cached) {
      return res.json(cached);
    }

    const {
      city,
      state,
      q,
      category,
      status = 'open',
      minValue,
      maxValue,
      minEmd,
      maxEmd,
      source,
      mseStartupOnly,
      zeroExperienceOnly,
      sort = 'endDate_asc',
      page = '1',
      limit = '20',
    } = req.query;

    const where = {};

    if (city && city !== 'all') where.locationCity = city;
    if (state && state !== 'all') where.locationState = state;
    if (source && source !== 'all') where.source = source.toUpperCase();
    if (status && status !== 'all') where.status = status;

    if (q) {
      where.OR = [
        { title: { contains: q, mode: 'insensitive' } },
        { organization: { contains: q, mode: 'insensitive' } },
        { bidNumber: { contains: q, mode: 'insensitive' } },
      ];
    }

    if (category) {
      const cats = Array.isArray(category) ? category : String(category).split(',');
      where.category = { hasSome: cats };
    }

    if (minValue || maxValue) {
      where.bidValue = {};
      if (minValue) where.bidValue.gte = Number(minValue);
      if (maxValue) where.bidValue.lte = Number(maxValue);
    }

    if (minEmd || maxEmd) {
      where.emdAmount = {};
      if (minEmd) where.emdAmount.gte = Number(minEmd);
      if (maxEmd) where.emdAmount.lte = Number(maxEmd);
    }

    const pageNum  = Math.max(1, parseInt(page, 10)  || 1);
    const limitNum = Math.min(100, Math.max(1, parseInt(limit, 10) || 20));

    // ── MSE / Startup / Zero-Experience filters ──────────────────────────────
    // These used to load ALL records into JS memory (OOM on 25k records).
    // Now they hit indexed Boolean columns in the DB — zero extra memory cost.
    if (mseStartupOnly === 'true') {
      // Match tenders where MSE OR Startup exemption is confirmed
      const mseOr = [];
      mseOr.push({ mseExemption:     true });
      mseOr.push({ startupExemption: true });
      // Merge with any existing OR clauses (e.g. from keyword search)
      if (where.OR) {
        where.AND = [{ OR: where.OR }, { OR: mseOr }];
        delete where.OR;
      } else {
        where.OR = mseOr;
      }
    }

    if (zeroExperienceOnly === 'true') {
      where.yearsOfExperienceZero = true;
    }

    let tenders, total;

    // Standard database pagination — handles all filter combinations
    {
      let orderBy = SORT_MAP[sort] || SORT_MAP.endDate_asc;

      if (sort === 'endDate_asc' || !sort) {
        const whereWithDate = { ...where, endDate: { not: null } };
        const whereNullDate = { ...where, endDate: null };

        total = await prisma.tender.count({ where });
        const withDate = await prisma.tender.findMany({
          where: whereWithDate,
          orderBy: { endDate: 'asc' },
          skip: (pageNum - 1) * limitNum,
          take: limitNum,
        });

        let remaining = limitNum - withDate.length;
        let nullDateItems = [];
        if (remaining > 0) {
          const dateCount = await prisma.tender.count({ where: whereWithDate });
          const skipNull  = Math.max(0, (pageNum - 1) * limitNum - dateCount);
          nullDateItems = await prisma.tender.findMany({
            where: whereNullDate,
            orderBy: { fetchedAt: 'desc' },
            skip: skipNull,
            take: remaining,
          });
        }
        tenders = [...withDate, ...nullDateItems];
      } else {
        total   = await prisma.tender.count({ where });
        tenders = await prisma.tender.findMany({
          where,
          orderBy,
          skip: (pageNum - 1) * limitNum,
          take: limitNum,
        });
      }
    }

    const result = { tenders, total, page: pageNum, limit: limitNum };
    cacheSet(cacheKey, result, 3600000); // 1 hour

    res.json(result);
  } catch (e) {
    console.error('[api] GET /tenders error:', e);
    res.status(500).json({ error: 'Internal server error' });
  }
});

/** GET /api/tenders/:source/:bidNumber — single tender detail (full record) */
router.get('/:source/:bidNumber', async (req, res) => {
  try {
    const { source, bidNumber } = req.params;
    const tender = await prisma.tender.findUnique({
      where: { source_bidNumber: { source: source.toUpperCase(), bidNumber } },
    });
    if (!tender) return res.status(404).json({ error: 'Tender not found' });
    res.json(tender);
  } catch (e) {
    console.error('[api] GET /tenders/:source/:bidNumber error:', e);
    res.status(500).json({ error: 'Internal server error' });
  }
});

/** GET /api/tenders/:source/:bidNumber/document — stream saved PDF */
router.get('/:source/:bidNumber/document', async (req, res) => {
  try {
    const { source, bidNumber } = req.params;
    const tender = await prisma.tender.findUnique({
      where: { source_bidNumber: { source: source.toUpperCase(), bidNumber } },
    });
    if (!tender || !tender.pdfPath || !fs.existsSync(tender.pdfPath)) {
      return res.status(404).json({ error: 'Document not available' });
    }
    res.setHeader('Content-Type', 'application/pdf');
    res.setHeader('Content-Disposition', `inline; filename="${source}-${bidNumber}.pdf"`);
    fs.createReadStream(tender.pdfPath).pipe(res);
  } catch (e) {
    console.error('[api] GET /tenders/:source/:bidNumber/document error:', e);
    res.status(500).json({ error: 'Internal server error' });
  }
});

/**
 * GET /api/tenders/:source/:bidNumber/ai-extract
 *
 * Returns the AI-structured extraction (consignees, eligibility, atc) for a GEM tender.
 *
 * Strategy (Option C — eager pipeline + on-demand fallback):
 *   1. If sourceMeta.aiExtract already stored in DB → return instantly.
 *   2. If the tender has a saved PDF but no aiExtract → run AI live,
 *      persist to DB, then return.
 *   3. No PDF and no stored data → 404 with explanation.
 *
 * Only available for GEM tenders (CSPGCL uses the built-in regex parser).
 */
router.get('/:source/:bidNumber/ai-extract', async (req, res) => {
  try {
    const { source, bidNumber } = req.params;

    if (source.toUpperCase() !== 'GEM') {
      return res.status(400).json({
        error: 'AI extraction is only available for GEM tenders',
        hint: 'CSPGCL tenders use the built-in table parser',
      });
    }

    const tender = await prisma.tender.findUnique({
      where: { source_bidNumber: { source: 'GEM', bidNumber } },
    });
    if (!tender) return res.status(404).json({ error: 'Tender not found' });

    // ── 1. Serve stored DB result (set by pipeline) ─────────────────────────
    const stored = tender.sourceMeta?.aiExtract;
    if (stored && stored.extractedAt) {
      return res.json({
        source:    'db_cache',
        bidNumber,
        title:     tender.title,
        ...stored,
      });
    }

    // ── 2. On-demand: run AI if PDF is present ──────────────────────────────
    if (!tender.pdfPath || !fs.existsSync(tender.pdfPath)) {
      return res.status(404).json({
        error:  'AI extract not yet available',
        reason: 'No PDF downloaded for this tender yet. Wait for the next pipeline run.',
      });
    }

    console.log(`[api] on-demand AI extraction for ${bidNumber}...`);

    let rawText = '';
    try {
      const buf  = fs.readFileSync(tender.pdfPath);
      const data = await pdfParse(buf);
      rawText    = data.text || '';
    } catch (e) {
      console.error('[api] pdf-parse failed for on-demand extract:', e.message);
      return res.status(500).json({ error: 'Failed to parse PDF', detail: e.message });
    }

    const aiResult = await aiExtractTender(rawText);
    if (!aiResult) {
      return res.status(503).json({
        error:  'AI extraction failed',
        reason: 'OpenRouter call failed or AI_EXTRACT_ENABLED=false. Check server logs.',
      });
    }

    // Persist to DB so next request is instant (no repeat AI call)
    const updatedSourceMeta = {
      ...(tender.sourceMeta || {}),
      aiExtract: aiResult,
    };
    await prisma.tender.update({
      where: { id: tender.id },
      data:  { sourceMeta: updatedSourceMeta },
    });

    console.log(`[api] ✓ on-demand AI extract stored to DB for ${bidNumber}`);

    return res.json({
      source:    'live',
      bidNumber,
      title:     tender.title,
      ...aiResult,
    });
  } catch (e) {
    console.error('[api] GET /tenders/:source/:bidNumber/ai-extract error:', e);
    res.status(500).json({ error: 'Internal server error' });
  }
});

export default router;

File: backend/app/routes/tenders.py
"""
app/routes/tenders.py

Direct port of src/routes/tenders.js.

NOTE: the `/ai-extract` route's on-demand AI extraction depended on the
GeM offline extractor (src/extractor/*.js), which was not converted per
instruction ("pdf extractor not needed"). It now returns the stored
sourceMeta.aiExtract if present (same as before), otherwise a 404
explaining extraction isn't available — same shape as the original
"no PDF yet" branch, just without the on-demand fallback.
"""
import json
import os
import re
from datetime import datetime, timezone

from fastapi import APIRouter, Request, Query
from fastapi.responses import JSONResponse, FileResponse

from .. import db
from ..cache import get as cache_get, set as cache_set
from ..serialize import serialize_doc

router = APIRouter(prefix="/api/tenders", tags=["tenders"])

SORT_MAP = {
    'endDate_asc': [('endDate', 1)],
    'endDate_desc': [('endDate', -1)],
    'bidValue_asc': [('bidValue', 1)],
    'bidValue_desc': [('bidValue', -1)],
    'emdAmount_asc': [('emdAmount', 1)],
    'emdAmount_desc': [('emdAmount', -1)],
    'fetchedAt_desc': [('fetchedAt', -1)],
}


@router.get("")
@router.get("/")
async def list_tenders(
    request: Request,
    city: str = None,
    state: str = None,
    q: str = None,
    category: str = None,
    status: str = 'open',
    minValue: float = None,
    maxValue: float = None,
    minEmd: float = None,
    maxEmd: float = None,
    source: str = None,
    plant: str = None,
    mseStartupOnly: str = None,
    zeroExperienceOnly: str = None,
    sort: str = 'endDate_asc',
    page: int = 1,
    limit: int = 20,
):
    try:
        cache_key = 'tenders:' + json.dumps(dict(request.query_params), sort_keys=True)
        cached = cache_get(cache_key)
        if cached:
            return JSONResponse(cached)

        where = {}
        and_conditions = []

        if city and city != 'all':
            where['locationCity'] = city
        if state and state != 'all':
            where['locationState'] = state
        if source and source != 'all':
            where['source'] = source.upper()
        if plant and plant != 'all':
            where['plantId'] = plant

        now = datetime.now(timezone.utc)
        if status == 'open':
            where['status'] = 'open'
            and_conditions.append({'$or': [{'endDate': {'$gte': now}}, {'endDate': None}]})
        elif status and status != 'all':
            where['status'] = status

        if q:
            escaped = re.escape(q)
            and_conditions.append({'$or': [
                {'title': {'$regex': escaped, '$options': 'i'}},
                {'organization': {'$regex': escaped, '$options': 'i'}},
                {'bidNumber': {'$regex': escaped, '$options': 'i'}},
            ]})

        if category:
            cats = category.split(',') if isinstance(category, str) else category
            where['category'] = {'$in': cats}

        if minValue is not None or maxValue is not None:
            rng = {}
            if minValue is not None:
                rng['$gte'] = minValue
            if maxValue is not None:
                rng['$lte'] = maxValue
            where['bidValue'] = rng

        if minEmd is not None or maxEmd is not None:
            rng = {}
            if minEmd is not None:
                rng['$gte'] = minEmd
            if maxEmd is not None:
                rng['$lte'] = maxEmd
            where['emdAmount'] = rng

        page_num = max(1, page)
        limit_num = min(10000, max(1, limit))

        if mseStartupOnly == 'true':
            and_conditions.append({'$or': [{'mseExemption': True}, {'startupExemption': True}]})

        if zeroExperienceOnly == 'true':
            where['yearsOfExperienceZero'] = True

        if and_conditions:
            where['$and'] = and_conditions

        col = db.tenders

        if sort == 'endDate_asc' or not sort:
            where_with_date = {**where, 'endDate': {'$ne': None}}
            where_null_date = {**where, 'endDate': None}

            total = await col.count_documents(where)
            with_date = await col.find(where_with_date).sort('endDate', 1).skip((page_num - 1) * limit_num).limit(limit_num).to_list(length=limit_num)

            remaining = limit_num - len(with_date)
            null_date_items = []
            if remaining > 0:
                date_count = await col.count_documents(where_with_date)
                skip_null = max(0, (page_num - 1) * limit_num - date_count)
                null_date_items = await col.find(where_null_date).sort('fetchedAt', -1).skip(skip_null).limit(remaining).to_list(length=remaining)

            tenders = with_date + null_date_items
        else:
            order_by = SORT_MAP.get(sort, SORT_MAP['endDate_asc'])
            total = await col.count_documents(where)
            tenders = await col.find(where).sort(order_by).skip((page_num - 1) * limit_num).limit(limit_num).to_list(length=limit_num)

        result = {
            'tenders': [serialize_doc(t) for t in tenders],
            'total': total,
            'page': page_num,
            'limit': limit_num,
        }
        cache_set(cache_key, result, 3600000)  # 1 hour

        return JSONResponse(result)
    except Exception as e:
        print(f"[api] GET /tenders error: {e}")
        return JSONResponse({'error': 'Internal server error'}, status_code=500)


@router.get("/{source}/{bid_number}")
async def get_tender(source: str, bid_number: str):
    """GET /api/tenders/:source/:bidNumber — single tender detail (full record)"""
    try:
        tender = await db.tenders.find_one({'source': source.upper(), 'bidNumber': bid_number})
        if not tender:
            return JSONResponse({'error': 'Tender not found'}, status_code=404)
        return JSONResponse(serialize_doc(tender))
    except Exception as e:
        print(f"[api] GET /tenders/:source/:bidNumber error: {e}")
        return JSONResponse({'error': 'Internal server error'}, status_code=500)


@router.get("/{source}/{bid_number}/document")
async def get_tender_document(source: str, bid_number: str):
    """GET /api/tenders/:source/:bidNumber/document — stream saved PDF"""
    try:
        tender = await db.tenders.find_one({'source': source.upper(), 'bidNumber': bid_number})
        if not tender or not tender.get('pdfPath') or not os.path.exists(tender['pdfPath']):
            return JSONResponse({'error': 'Document not available'}, status_code=404)
        return FileResponse(
            tender['pdfPath'],
            media_type='application/pdf',
            headers={'Content-Disposition': f'inline; filename="{source}-{bid_number}.pdf"'},
        )
    except Exception as e:
        print(f"[api] GET /tenders/:source/:bidNumber/document error: {e}")
        return JSONResponse({'error': 'Internal server error'}, status_code=500)


@router.get("/{source}/{bid_number}/ai-extract")
async def get_ai_extract(source: str, bid_number: str):
    """GET /api/tenders/:source/:bidNumber/ai-extract

    Returns the stored AI-structured extraction (consignees, eligibility, atc)
    for a GEM tender, if the DB already has one cached from a prior run.
    On-demand live extraction is unavailable in this build (see module docstring).
    """
    try:
        if source.upper() != 'GEM':
            return JSONResponse({
                'error': 'AI extraction is only available for GEM tenders',
                'hint': 'CSPGCL tenders use the built-in table parser',
            }, status_code=400)

        tender = await db.tenders.find_one({'source': 'GEM', 'bidNumber': bid_number})
        if not tender:
            return JSONResponse({'error': 'Tender not found'}, status_code=404)

        stored = (tender.get('sourceMeta') or {}).get('aiExtract')
        if stored and stored.get('extractedAt'):
            return JSONResponse(serialize_doc({
                'source': 'db_cache',
                'bidNumber': bid_number,
                'title': tender.get('title'),
                **stored,
            }))

        return JSONResponse({
            'error': 'AI extract not yet available',
            'reason': 'No stored extraction for this tender, and on-demand extraction is not enabled in this deployment.',
        }, status_code=404)
    except Exception as e:
        print(f"[api] GET /tenders/:source/:bidNumber/ai-extract error: {e}")
        return JSONResponse({'error': 'Internal server error'}, status_code=500)
